man/geolookup.1
.\" Copyright (c) 2024, Peter Haag
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\"  * Redistributions of source code must retain the above copyright notice,
.\"    this list of conditions and the following disclaimer.
.\"  * Redistributions in binary form must reproduce the above copyright notice,
.\"    this list of conditions and the following disclaimer in the documentation
.\"    and/or other materials provided with the distribution.
.\"  * Neither the name of the author nor the names of its contributors may be
.\"    used to endorse or promote products derived from this software without
.\"    specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate$
.Dt GEOLOOKUP 1
.Os
.Sh NAME
.Nm geolookup
.Nd geo location lookup for IP addresses
.Sh SYNOPSIS
.Nm 
.Op Fl G Ar geoDBfile
.Op Fl h
.Ar iplist
.Nm
.Fl d Ar directory
.Fl w Ar geoDBfile
.Sh DESCRIPTION
.Nm 
is a tool to lookup AS and geo location information of one or more IP addresses. You need
to create the nfdump specific lookup database first, before using
.Nm
.Pp
.Nm
is also used to create the nfdump formatted lookup database file from the maxmind csv files.
You need to have a maxmind account at https://maxmind.com, in order to download the relevant
csv file.  See the section below for the building instructions.
.Pp
.Nm
accepts a list of IP addresses either on the command line, separated by spaces
or on
.Ar stdin
line by line. The IP address on each line can be embedded in a string separated by
spaces on the left and right, therefore it can read the piped output from other tools.
.Pp
.Nm
accepts a list of AS numbers prefixed with
.Sy as
and separated with spaces
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d Ar directory
Use the csv files in 
.Ar directory
to build the binary lookup database file. With this argument
.Nm
creates a new binary lookup database file.
.It Fl w Ar geoDBfile
Name of the new lookup database file.
.It Fl G Ar geoDBfile
Use the binary geoDBfile as lookup database for the current AS and location lookups.
.It Fl h
Print help and exit.
.El
.Pp
To specify the geo lookup database 
.Nm
searches at the following places:
.Bl -bullet -compact
.It
If the default
.Ar nfdump(1)
configuration file exists, it reads the tag
.Ar geodb.path
.It
If the environment variable
.Ar NFGEODB
is set, use this setting as lookup database.
.It
The command line argument
.Fl G
.El
If multiple locations are given, the environment variable
.Ar NFGEODB
overwrites the nfdump config file and the command line option
.Fl G
overwrites the environment variable
.Ar NFGEODB .
.Sh ENVIRONMENT
.Nm
reads the environment variable
.Ar NFGEODB
.Sh EXIT STATUS
.Nm
returns 0 on success and 255 otherwise.
.Sh EXAMPLES
To create a lookup database, you need an account at
.Ar https://maxmind.com.
It works with the paid GeoDB as well as with the free GeoLite2 csv files. The easiest way for creating or 
updating the binary lookup database is the use of the script
.Ar updateGeoDB.sh
provided with all other nfdump files. Insert your license key and run the script. This creates a new lookup database.
.Pp
If you want to do it manually follow these steps:
.Bl -enum -compact
.It
Log into https://maxmind.com and download the corresponding csv files: GeoLite2-ASN-CSV and GeoLite2-City-CSV
.It
Unpack the zip files and put the following csv files into a newly created build directory:
GeoLite2-ASN-Blocks-IPv4.csv, GeoLite2-ASN-Blocks-IPv6.csv, GeoLite2-City-Blocks-IPv4.csv, GeoLite2-City-Blocks-IPv6.csv, GeoLite2-City-Locations-en.csv
Maxmind offers several languages for the City-Locations file. Choose only one.
.It
.Nm
.Fl d Ar builddir Fl w Ar geofile
.It 
Move the lookup database to the final location.
.El
.Sh SEE ALSO
.Ar nfdump
has already builtin lookup options to decorate the text output with geo location and AS information.
.Pp
.Xr nfdump 1


ChangeLog
- 972d1fb 2026-04-28  (HEAD -> devel_1.8.x, origin/devel_1.8.x) Update Readme.md
- 8fa4fee 2026-04-28  Update Readme.md
- 50d99c6 2026-04-28  Refine and improve test suite
- df9c1a8 2026-04-28  Modify testing
- a4e49cf 2026-04-28  Cleanup includes
- 3118e88 2026-04-28  Remove compat_1.6 files
- 448166e 2026-04-27  Adapt nfreader to nffileV3 and recordV4
- 1e8a820 2026-04-27  Fix RenameAppend()
- 6b913f7 2026-04-27  Make ident fixlen and adapt tests
- 3c867c0 2026-04-26  Fix some test files
- b7d60bb 2026-04-26  Fix nfreplay bug and add nfcapd test script
- 7431904 2026-04-25  Add tor node role. Still report only exit nodes
- bd43350 2026-04-25  Fix comments
- f0d61d2 2026-04-25  Update tor code with IPv6 exit nodes
- 36a555c 2026-04-25  Cleanup torlookup code
- e50dd3f 2026-04-25  Review maxmind code and update minor issues
- 5a3ae2b 2026-04-24  Improve -G/-H handling with .flat cache file.. Add flatpath option in config file
- 4cb03fc 2026-04-24  Fix payload size in convertV2
- 396a00b 2026-04-24  Improve maxmind lookup - add flat cache
- 1ee2ff3 2026-04-24  Improve tor lookup
- 62eda08 2026-04-24  Cleanup maxmind geolookup and tor torlookup
- ec67890 2026-04-24  Cleanup lauchner and repeater
- 43a3353 2026-04-23  Adapt and improve  nflowcache.c for aggregation to v4 record
- 2fcfa54 2026-04-23  Cleanup custom aggregation for new v4 record
- 9e63106 2026-04-22  Re-enable ChangeIdent
- 1642d9c 2026-04-22  Re-enable sat_record printing
- af0d1a2 2026-04-22  Fix spelling errors
- 4827cd4 2026-04-22  Fix minor bugs
- a985f62 2026-04-11  Work on more fixes. Add entropy to payload printing in output_raw
- d5ced32 2026-04-10  Work on XXX issues
- 9f3af9a 2026-04-10  Work on filter
- 0bc6c33 2026-04-10  Adapt legacy programs to new layout
- d023b74 2026-04-08  Fixes from debug session
- 73e5bf2 2026-04-08  Add nbar array records and if/vrf records
- 276d231 2026-04-08  Free converter exporter tables
- cbf5c87 2026-04-07  Fix some bugs in exporter conversion and some cleanup
- 41e8989 2026-04-07  Add v2 exporter/sampler info to V2 converter
- e190f50 2026-04-06  Add v3 converter and fix v1/v5 collector
- cfebf09 2026-04-05  Fix bug and add stub  for 1.7.x conversion
- 4a80f2e 2026-04-05  Add missing new files
- d3aa73c 2026-04-05  Implemented nffile version 3
- fc6c5ff 2026-03-30  Enable extensions in grammar.y
- 0cd2fb1 2026-03-30  Replace all gettimeofday by clock_gettime. cleanup nfxv3.h includes
- a91643f 2026-03-30  Update all programs to v4 record
- 50595b7 2026-03-29  Update sfcapd to new v4 record
- 9fd446d 2026-03-29  - Pipeline: replace rank offset with extension offset. Fill offset table at L_END and not at start. - Netflow v5. Check at runtime to add optional extension. Prevent wasting space.
- b3ad103 2026-03-28  Status 20260328-1
- 24daa81 2026-03-26  Status 20260327-1
- d9504e5 2026-03-24  Switch to devel_1.8.x branch. Add new record v4
- 6b96df6 2026-03-04  (origin/hackathon2026) Fix proper EXnull handling and printing and update tests
- 5927ed3 2026-03-04  Fix bugs
- b29f71d 2026-03-03  Replace old number scaling code
- ba94176 2026-03-02  Avoid expensive stat() calls. Replaced by fstatat()
- ff26f32 2026-03-02  New code base for nfexpire
- 732b1e4 2026-02-28  Refrag code - move logging functions into new file logging.[ch]
- 1596835 2026-02-26  Fix other thread unsafe glibc library calls
- 5b379bf 2026-02-26  Replace all localtime() calls with localtime_r()
- e36b865 2026-02-26  Fix nfprofile and some localtime() issues
- 28b42e3 2026-02-25  Fix SetupPath()
- a59e770 2026-02-25  New bookkeeper implemented
- aac815d 2026-02-25  Remove old expire code in collector
- 13f63b0 2026-02-24  Update sflow collector to nfcapd level
- ab39ba7 2026-02-23  Updated launcher code in nfcapd
- d750f97 2026-02-23  Ready new repeater code in nfcapd
- f4c9f53 2026-02-20  Replaced run() loop in nfcapd
- d4a19a6 2026-02-17  split backend data in new nffile backend context
- 4d81212 2026-02-16  Remove old and no longer used libpcap life code
- 6c6c87c 2026-02-16  Fix #667 duration bug in -s stat output
- 14d0b46 2026-02-16  Cleanup TPACKETV3 code
- 84f3b22 2026-02-16  Cleanup BPF code. Make buffer size configurable
- fce61d7 2026-02-16  Fix IP fragment init code
- 361bd5e 2026-02-15  Fix Statistic on BPF socket
- 09de43b 2026-02-15  Fix nfpcapd build options and remove useless FIX_INCLUDES
- dbdc2ba 2026-02-14  Fix potential swap32() double definition
- b3e7370 2026-02-14  Add expire function to fragment list in nfpcapd
- 8dfa9e5 2026-02-14  Update Readme
- b9438dd 2026-02-14  Update nfpcapd - Refactor messy ProcessPacket code with state processing
- e7af66f 2026-02-12  Update nfpcapd - Improve packet processing. Make it more robust
- 9d585f8 2026-02-10  Update debug messages
- 8c29f61 2026-02-10  Replace verbose option -v to verbose level -v <n> for most binaries. Needs more work
- 3ea8d71 2026-02-09  Update nfpcapd - remove unused interrupt handler and use atomic done variables
- 9a14047 2026-02-09  Fix statistic for pcap file processing
- a2b1eb7 2026-02-09  Update nfpcapd - replace slow packet dump code by native calls
- 5ab715f 2026-02-05  Add missing packet_pcap_file.c
- 183f045 2026-02-05  Update nfpcapd - Replace pcap reader code with faster mmap version
- ac07fa1 2026-02-04  Silently skip unknown records in nfanon. Fixed #665
- 5aedf78 2026-02-04  Update nfpcapd - Modify slab allocator
- 3490191 2026-02-03  Update nfpcapd - replace rb-tree with an open hash. Improves nfpcapd performance
- dcabdb6 2026-02-01  Update nfpcapd - shrink slab check
- 0cebb7c 2026-02-01  Update nfpcapd - Fix verbose logging
- 8360a58 2026-02-01  Update nfpcapd - Update help text and man page for node cache size
- 95b252e 2026-02-01  Update nfpcapd - Replace ever growing node cache with an adaptive slab allocator
- edd6784 2026-01-31  Fix #662. Update man page
- 29cf020 2026-01-29  Update nfpcapd - Rework a lot of code. Remove lock for NewNode() in hot path

src/maxmind/geolookup.c
/*
 *  Copyright (c) 2024, Peter Haag
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *
 *   * Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *     and/or other materials provided with the distribution.
 *   * Neither the name of the author nor the names of its contributors may be
 *     used to endorse or promote products derived from this software without
 *     specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <arpa/inet.h>
#include <ctype.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include "conf/nfconf.h"
#include "maxmind/maxmind.h"
#include "maxmind/mmhash.h"
#include "mmcreate.h"
#include "nffile.h"
#include "util.h"

static void usage(char *name) {
    printf(
        "usage %s [options] \n"
        "-h\t\tthis text you see right here.\n"
        "-G <dir>\tmaxmind GeoDB in nfdump format to lookup info.\n"
        "-d <dir>\tDirectory containing the maxmind .csv files to convert into nfdump GeoDB.\n"
        "-w <file>\tName of nfdump GeoDB file.\n",
        name);
} /* usage */

// Return a pointer to the trimmed string
static char *string_trim(char *s) {
    while (isspace((unsigned char)*s)) s++;
    if (*s) {
        char *p = s;
        while (*p) p++;
        while (isspace((unsigned char)*(--p)));
        p[1] = '\0';
    }

    return s;
}  // end of string_trim

static int valid_ipv4(char *s) {
    char *c = s;
    int i = 0;
    while (*c) {
        if ((!isdigit((unsigned char)*c) && *c != '.') || i > 15) {
            return 0;
        }
        c++;
        i++;
    }

    c = strdup(s);
    int numbers = 0;
    char *sep = ".";
    char *brkt;
    char *ns = strtok_r(c, sep, &brkt);
    while (ns) {
        int num = atoi(ns);
        if (num > 255) {
            free(c);
            return 0;
        }
        numbers++;
        ns = strtok_r(NULL, sep, &brkt);
    }

    free(c);
    return numbers == 4;
}

static int valid_ipv6(char *s) {
    char *c = s;
    int i = 0;
    while (*c) {
        if ((!isxdigit((unsigned char)*c) && *c != ':' && *c != '.') || i > 39) {
            return 0;
        }
        c++;  // point to next character
    }
    if (strchr(s, ':') == NULL) {
        return 0;
    }
    c = strdup(s);
    char *brkt;
    char *ns = strtok_r(c, ":", &brkt);
    while (ns) {
        int num = atoi(ns);
        if (num > 65535) {
            free(c);
            return 0;
        }
        ns = strtok_r(NULL, ":", &brkt);
    }

    free(c);
    uint64_t u[2];
    if (inet_pton(PF_INET6, s, u) != 1) {
        return 0;
    }

    return 1;
}

int main(int argc, char **argv) {
    char *dirName = NULL;
    char *geoFile = getenv("NFGEODB");
    char *wfile = "mmc.nf";
    int c;
    while ((c = getopt(argc, argv, "hd:G:w:")) != EOF) {
        switch (c) {
            case 'h':
                usage(argv[0]);
                exit(0);
                break;
            case 'd':
                if (!CheckPath(optarg, S_IFDIR)) exit(EXIT_FAILURE);
                dirName = strdup(optarg);
                break;
            case 'w':
                wfile = optarg;
                break;
            case 'G':
                if (!CheckPath(optarg, S_IFREG)) exit(EXIT_FAILURE);
                geoFile = strdup(optarg);
                break;
            default:
                usage(argv[0]);
                exit(0);
        }
    }

    if (!Init_nffile(1, NULL)) exit(EXIT_FAILURE);

    if (dirName && wfile) {
        if (Init_MaxMind() == 0 || LoadMaps(dirName) == 0 || SaveMaxMind(wfile) == 0) {
            exit(EXIT_FAILURE);
        }
        exit(EXIT_SUCCESS);
    }

    if (geoFile == NULL) {
        if (ConfOpen(NULL, "nfdump") < 0) exit(EXIT_FAILURE);
        geoFile = ConfGetString("geodb.path");
    }

    if (geoFile == NULL) {
        LogError("Missing nfdump geo DB. -G or NFGEODB env required");
        exit(EXIT_FAILURE);
    }

    if (!LoadMaxMind(geoFile)) {
        LogError("Failed to load nfdump geo DB");
        exit(EXIT_FAILURE);
    }

    if (argc - optind > 0) {
        while (argc - optind > 0) {
            char *arg = argv[optind++];
            if (strlen(arg) > 2) {
                if (arg[0] == 'a' && arg[1] == 's') {
                    // do as Lookup
                    LookupAS(arg + 2);
                } else if ((valid_ipv4(arg) || valid_ipv6(arg))) {
                    LookupWhois(arg);
                } else {
                    LogError("Not a valid IPv4 or IPv6: %s", arg);
                    exit(EXIT_FAILURE);
                }
            }
        }
    } else {
        char *line = NULL;
        size_t linecap = 0;
        ssize_t lineLen;
        // read each line - trimm \n
        while ((lineLen = getline(&line, &linecap, stdin)) > 0) {
            if (lineLen > 1024) {
                LogError("Line length error");
                exit(EXIT_FAILURE);
            }
            // last line on stdin may not be newline terminated
            char *eol = strchr(line, '\n');
            if (eol) *eol = '\0';

            // split ' ' separated words and check, if it's an IPv4/v6
            char *sep = " (";
            char *word, *brkt;
            word = strtok_r(line, sep, &brkt);
            while (word) {
                if (valid_ipv4(word) || valid_ipv6(word)) {
                    LookupWhois(string_trim(word));
                }
                word = strtok_r(NULL, sep, &brkt);
            }
        }
    }

    return 0;
}
